File: backend/routes/notes.py
from flask import Blueprint, request, jsonify, redirect
from werkzeug.utils import secure_filename
from datetime import datetime
import cloudinary
import cloudinary.uploader
from models.note import Note
from models.user import User
from config.db import Database
from middleware.auth import auth_required, admin_required

notes_bp = Blueprint('notes', __name__)

@notes_bp.route('/upload', methods=['POST'])
@auth_required
def upload_note():
    try:
        title = request.form.get('title')
        subject = request.form.get('subject')
        description = request.form.get('description')
        file = request.files.get('file')

        if not file:
            return jsonify({'message': 'No file uploaded'}), 400

        # Upload to Cloudinary with raw resource type for PDF files
        try:
            # Get file extension
            file_ext = file.filename.split('.')[-1] if '.' in file.filename else 'pdf'
            public_id = f"{int(datetime.utcnow().timestamp())}_{file.filename}"
            
            result = cloudinary.uploader.upload(
                file,
                resource_type='raw',
                folder='notes-sharing',
                type='upload',
                public_id=public_id
            )
            print(f'Cloudinary upload result: {result}')
            print(f'Secure URL: {repr(result.get("secure_url"))}')
        except Exception as cloudinary_error:
            print(f'Cloudinary Upload Error: {cloudinary_error}')
            return jsonify({'message': f'Cloudinary upload failed: {str(cloudinary_error)}'}), 500

        db = Database.get_db()
        note_data = {
            'title': title,
            'subject': subject,
            'description': description,
            'fileUrl': result['secure_url'],
            'fileName': file.filename,
            'uploadedBy': request.user['_id']
        }

        result = Note.create(db, note_data)
        note = Note.find_by_id(db, result.inserted_id)
        note['_id'] = str(note['_id'])
        note['uploadedBy'] = str(note['uploadedBy'])

        return jsonify(note), 201
    except Exception as e:
        return jsonify({'message': str(e)}), 500

@notes_bp.route('/', methods=['GET'])
def get_notes():
    try:
        search = request.args.get('search')
        subject = request.args.get('subject')

        db = Database.get_db()
        notes = Note.find_all(db, search, subject)

        # Populate user info
        for note in notes:
            user = User.find_by_id(db, note['uploadedBy'])
            if user:
                note['uploadedBy'] = {
                    'id': str(user['_id']),
                    'name': user['name'],
                    'email': user['email']
                }

        return jsonify(notes)
    except Exception as e:
        return jsonify({'message': str(e)}), 500

@notes_bp.route('/<note_id>', methods=['GET'])
def get_note(note_id):
    try:
        db = Database.get_db()
        note = Note.find_by_id(db, note_id)
        
        if not note:
            return jsonify({'message': 'Note not found'}), 404

        note['_id'] = str(note['_id'])
        note['uploadedBy'] = str(note['uploadedBy'])
        
        user = User.find_by_id(db, note['uploadedBy'])
        if user:
            note['uploadedBy'] = {
                'id': str(user['_id']),
                'name': user['name'],
                'email': user['email']
            }

        return jsonify(note)
    except Exception as e:
        return jsonify({'message': str(e)}), 500

@notes_bp.route('/<note_id>/download', methods=['GET'])
def download_note(note_id):
    try:
        print(f'\n=== DOWNLOAD REQUEST ===')
        print(f'note_id: {note_id}')
        db = Database.get_db()
        note = Note.find_by_id(db, note_id)
        
        if not note:
            print(f'Note not found: {note_id}')
            return jsonify({'message': 'Note not found'}), 404

        print(f'Note found: {note.get("title")}')
        print(f'Raw fileUrl from DB: {repr(note["fileUrl"])}')
        
        Note.increment_downloads(db, note_id)

        # Redirect to local file URL
        download_url = note['fileUrl']
        print(f'Redirecting to: {download_url}')
        print(f'====================\n')
        
        return redirect(download_url)
    except Exception as e:
        print(f'Download error: {e}')
        import traceback
        traceback.print_exc()
        return jsonify({'message': str(e)}), 500

@notes_bp.route('/<note_id>/like', methods=['POST'])
@auth_required
def like_note(note_id):
    try:
        db = Database.get_db()
        note = Note.find_by_id(db, note_id)
        
        if not note:
            return jsonify({'message': 'Note not found'}), 404

        user_id = str(request.user['_id'])
        liked_by = note.get('likedBy', [])
        
        # Convert to ObjectId for comparison
        from bson import ObjectId
        liked_by_oids = [ObjectId(uid) if isinstance(uid, str) else uid for uid in liked_by]
        user_oid = ObjectId(user_id)
        
        has_liked = user_oid in liked_by_oids

        if has_liked:
            new_likes = Note.update_likes(db, note_id, user_id, increment=False)
            liked = False
        else:
            new_likes = Note.update_likes(db, note_id, user_id, increment=True)
            liked = True

        return jsonify({'likes': new_likes, 'liked': liked})
    except Exception as e:
        return jsonify({'message': str(e)}), 500

@notes_bp.route('/<note_id>', methods=['DELETE'])
@auth_required
def delete_note(note_id):
    try:
        db = Database.get_db()
        note = Note.find_by_id(db, note_id)
        
        if not note:
            return jsonify({'message': 'Note not found'}), 404

        # Check if user is owner or admin
        if str(note['uploadedBy']) != str(request.user['_id']) and request.user.get('role') != 'admin':
            return jsonify({'message': 'Not authorized to delete this note'}), 403

        # Delete from Cloudinary
        public_id = note['fileUrl'].split('/')[-1].split('.')[0]
        cloudinary.uploader.destroy(f'notes-sharing/{public_id}')

        Note.delete_by_id(db, note_id)

        return jsonify({'message': 'Note deleted successfully'})
    except Exception as e:
        return jsonify({'message': str(e)}), 500

@notes_bp.route('/user/notes', methods=['GET'])
@auth_required
def get_user_notes():
    try:
        db = Database.get_db()
        notes = Note.find_by_user(db, str(request.user['_id']))

        # Populate user info
        for note in notes:
            user = User.find_by_id(db, note['uploadedBy'])
            if user:
                note['uploadedBy'] = {
                    'id': str(user['_id']),
                    'name': user['name'],
                    'email': user['email']
                }

        return jsonify(notes)
    except Exception as e:
        return jsonify({'message': str(e)}), 500

File: backend/routes/notes.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const cloudinary = require('../config/cloudinary');
const Note = require('../models/Note');
const { auth, adminAuth } = require('../middleware/auth');

// Multer setup for memory storage
const storage = multer.memoryStorage();
const upload = multer({ storage });

// Upload note
router.post('/upload', auth, upload.single('file'), async (req, res) => {
  try {
    const { title, subject, description } = req.body;

    if (!req.file) {
      return res.status(400).json({ message: 'No file uploaded' });
    }

    // Upload to Cloudinary
    const result = await new Promise((resolve, reject) => {
      cloudinary.uploader.upload_stream(
        {
          resource_type: 'auto',
          folder: 'notes-sharing',
        },
        (error, result) => {
          if (error) reject(error);
          else resolve(result);
        }
      ).end(req.file.buffer);
    });

    const note = new Note({
      title,
      subject,
      description,
      fileUrl: result.secure_url,
      fileName: req.file.originalname,
      uploadedBy: req.user._id,
    });

    await note.save();

    res.status(201).json(note);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Get all notes with search and filter
router.get('/', async (req, res) => {
  try {
    const { search, subject } = req.query;
    const query = {};

    if (search) {
      query.$or = [
        { title: { $regex: search, $options: 'i' } },
        { description: { $regex: search, $options: 'i' } },
      ];
    }

    if (subject) {
      query.subject = { $regex: subject, $options: 'i' };
    }

    const notes = await Note.find(query)
      .populate('uploadedBy', 'name email')
      .sort({ createdAt: -1 });

    res.json(notes);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Get single note
router.get('/:id', async (req, res) => {
  try {
    const note = await Note.findById(req.params.id).populate('uploadedBy', 'name email');
    
    if (!note) {
      return res.status(404).json({ message: 'Note not found' });
    }

    res.json(note);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Download note (increment download count)
router.get('/:id/download', async (req, res) => {
  try {
    const note = await Note.findById(req.params.id);
    
    if (!note) {
      return res.status(404).json({ message: 'Note not found' });
    }

    note.downloads += 1;
    await note.save();

    // Stream file from Cloudinary
    const response = await fetch(note.fileUrl);
    if (!response.ok) {
      return res.status(500).json({ message: 'Failed to fetch file from Cloudinary' });
    }

    // Set headers for download
    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename="${note.fileName}"`);
    
    // Pipe the file stream
    response.body.pipe(res);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Like note
router.post('/:id/like', auth, async (req, res) => {
  try {
    const note = await Note.findById(req.params.id);
    
    if (!note) {
      return res.status(404).json({ message: 'Note not found' });
    }

    const hasLiked = note.likedBy.includes(req.user._id);

    if (hasLiked) {
      note.likedBy.pull(req.user._id);
      note.likes -= 1;
    } else {
      note.likedBy.push(req.user._id);
      note.likes += 1;
    }

    await note.save();

    res.json({ likes: note.likes, liked: !hasLiked });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Delete note
router.delete('/:id', auth, async (req, res) => {
  try {
    const note = await Note.findById(req.params.id);
    
    if (!note) {
      return res.status(404).json({ message: 'Note not found' });
    }

    // Check if user is owner or admin
    if (note.uploadedBy.toString() !== req.user._id.toString() && req.user.role !== 'admin') {
      return res.status(403).json({ message: 'Not authorized to delete this note' });
    }

    // Delete from Cloudinary
    const publicId = note.fileUrl.split('/').pop().split('.')[0];
    await cloudinary.uploader.destroy(`notes-sharing/${publicId}`);

    await Note.findByIdAndDelete(req.params.id);

    res.json({ message: 'Note deleted successfully' });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Get user's notes
router.get('/user/notes', auth, async (req, res) => {
  try {
    const notes = await Note.find({ uploadedBy: req.user._id })
      .populate('uploadedBy', 'name email')
      .sort({ createdAt: -1 });

    res.json(notes);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

module.exports = router;
